src/Microsoft.AspNetCore.SignalR.Core/StreamTracker.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.SignalR.Protocol;
+
+namespace Microsoft.AspNetCore.SignalR
+{
+    internal class StreamTracker
+    {
+        private static readonly MethodInfo _buildConverterMethod = typeof(StreamTracker).GetMethods().Single(m => m.Name.Equals("BuildStream"));
+        public ConcurrentDictionary<string, IStreamConverter> Lookup = new ConcurrentDictionary<string, IStreamConverter>();
+
+        /// <summary>
+        /// Creates a new stream and returns the ChannelReader for it as an object.
+        /// </summary>
+        public object AddStream(string streamId, Type itemType)
+        {
+            var newConverter = (IStreamConverter)_buildConverterMethod.MakeGenericMethod(itemType).Invoke(null, Array.Empty<object>());
+            Lookup[streamId] = newConverter;
+            return newConverter.GetReaderAsObject();
+        }
+
+        public Task ProcessItem(StreamItemMessage message)
+        {
+            return Lookup[message.InvocationId].WriteToStream(message.Item);
+        }
+
+        public void Complete(StreamCompleteMessage message)
+        {
+            Lookup.TryRemove(message.StreamId, out var converter);
+            converter.TryComplete(message.HasError ? new Exception(message.Error) : null);
+        }
+
+        public static IStreamConverter BuildStream<T>()
+        {
+            return new ChannelConverter<T>();
+        }
+
+        public interface IStreamConverter
```

Comment:
Will move to a new file, but the weird bit is that `StreamTracker.BuildStream` needs to be public, because of the `_buildConverterMethod` manuever. So it's return type `IStreamConverter` needs to public as well. Is there a good workaround?

Reply:
Andrew saved me.